packages/babel-plugin/src/utils/css-map.ts

```diff
@@ -0,0 +1,263 @@
+
+  if (!bindingIdentifier) return undefined;
+
+  // Evaluate the binding identifier to get the value of the CSS Map
+  const { value, meta: updatedMeta } = evaluateIdentifier(
+    bindingIdentifier,
+    meta,
+    evaluateExpression
+  );
+
+  // Ensure cssMap is used in a correct format.
+  if (
+    t.isCallExpression(value) &&
+    t.isIdentifier(value.callee) &&
+    value.callee.name === meta.state.compiledImports?.cssMap
+  ) {
+    // It's CSS Map! We now need to check a few things to ensure CSS Map is correctly used.
+    // We need to ensure cssMap receives only one argument.
+    if (value.arguments.length !== 1) {
+      throw buildCodeFrameError(
+        createErrorMessage(ErrorMessages.NUMBER_OF_ARGUMENT),
+        value,
+        updatedMeta.parentPath
+      );
+    }
+
+    // We need to ensure the argument is an objectExpression.
+    if (!t.isObjectExpression(value.arguments[0])) {
+      throw buildCodeFrameError(
+        createErrorMessage(ErrorMessages.ARGUMENT_TYPE),
+        value,
+        updatedMeta.parentPath
+      );
+    }
+
+    // We need to ensure callExpression isn't used to access a variant.
+    if (t.isCallExpression(expression.object)) {
+      throw buildCodeFrameError(
+        createErrorMessage(ErrorMessages.VARIANT_CALL_EXPRESSION),
+        expression.object,
+        meta.parentPath
+      );
+    }
+
+    // We disallows a nested CSS Map. e.g. { danger: { veryDanger: { ... } } }
+    if (t.isMemberExpression(expression.object)) {
+      throw buildCodeFrameError(
+        createErrorMessage(ErrorMessages.NESTED_VARIANT),
+        expression.object,
+        meta.parentPath
+      );
+    }
+
+    // We need to ensure the cssMap is accessed using a string literal or an identifier.
+    const property = getKeyOrProperty(expression);
+    if (!property) {
+      throw buildCodeFrameError(
+        createErrorMessage(ErrorMessages.VARIANT_ACCESS),
+        expression.property,
+        meta.parentPath
+      );
+    }
+
+    return {
+      value: value.arguments[0],
+      property,
+      meta: updatedMeta,
+    };
+  }
+
+  // It's not a CSS Map, let other code handle it
+  return undefined;
+};
+
+export const isCSSMap = (
+  expression: t.Expression,
+  meta: Metadata,
+  evaluateExpression: EvaluateExpression
+): boolean => {
+  return getCssMap(expression, meta, evaluateExpression) !== undefined;
+};
+
+/**
+ * Transform expression that uses a CSS Map into an array of logical expressions.
+ * For example:
+ * ```js
+ * const borderStyleMap = cssMap({
+ *   none: { borderStyle: 'none' },
+ *   solid: { borderStyle: 'solid' },
+ * });
+ * const Component = ({ borderStyle }) => <div css={css(borderStyleMap[borderStyle])} />
+ * ```
+ * gets transformed into:
+ * ```js
+ * const Component = ({ borderStyle }) => <div css={css([
+ *   borderStyle === 'none' && { borderStyle: 'none' },
+ *   borderStyle === 'solid' && { borderStyle: 'solid'}
+ * ])} />
```

Comment:
This looks like it's outputting an intermediate form. It's preferable for us to consolidate things into a 1-pass compilation rather than having multiple passes, because this is lossy and prevents us from understanding the original context (which is useful for optimisations like dropping the `"success"==="danger"` case. What's inhibiting us from doing this in the same pass as the `css`-prop -> `className` transform?

Reply:
um...good point. Let me look into it more.

Reply:
> consolidate things into a 1-pass compilation rather than having multiple passes

It's been architectural thing. 
- Compiled firstly evaluates user input, and compiles it to expression. e.g. `foo ? 'color: red': 'color: green'`
- At later stage, then compiles it to classNames and sheets

I think we can change this two-pass compilation into 1-pass by using local transform, as mentioned from another comment. I will make sure that I will spike this 1-pass compilation before PR is merged.